packages/sdk/src/oracle/services/statemachine.ts

```diff
@@ -0,0 +1,108 @@
+import assert from "assert";
+import type { Context, Handlers, Memory, Emit, ContextProps } from "../types/statemachine";
+import uid from "lodash/uniqueId";
+
+export { Context, Handlers, Emit };
+
+/**
+ * StateMachine. This class will be used to handle all change requests by the user, including setting state which
+ * may require triggering fetching data from chain, or initiating transactions that require tracking.
+ *
+ * This class is meant to step through states of a Context object. This object can have variable
+ * parameters and memory. There are several concepts to understand before using this class:
+ *
+ * 1. Handlers - This is the state machine state handlers, it is an objected keyed by each state, with a function handler.
+ * 2. Params - These are common parameters set by the caller, passed into each handler function.
+ * 3. Memory - This is a writing space within the state machine handlers that accumulates through states.
+ * 4. Context - This is the sum of all data needed to transition through the state machine states. It includes params, memory and metadata about the runtime.
```

Comment:
I try to build a mental model of this implementation. Is the Context smth comparable to the [Unit of Work Pattern](https://martinfowler.com/eaaCatalog/unitOfWork.html)? I see it as a db transaction

Reply:
thats really cool, thanks for linking to that, it is really similar. the difference from what i can see is that unit of work drafts changes to your state for execution, and then commits them, where as the design i have in mind is each state machine is just making changes to state in real time.  These could also be called "[commands](https://martinfowler.com/bliki/CommandQuerySeparation.html)" as the broader framework follows CQRS principles.

The idea is that all requests for change that originate on the user side are commands that execute within the statemachine. This allows us an environment where when executing the commands they can operate in a space with consistent state and safely issue side effects.  This implies though that all commands must execute serially to make guarantees for state consistency.

Reply:
this is also similar to "jobs" https://cloud.google.com/kubernetes-engine/docs/how-to/jobs